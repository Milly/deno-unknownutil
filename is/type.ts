/**
 * A type predicate function.
 */
export type Predicate<T> = (x: unknown) => x is T;

/**
 * A type predicated by Predicate<T>.
 *
 * ```ts
 * import { is, type PredicateType } from "https://deno.land/x/unknownutil@$MODULE_VERSION/mod.ts";
 *
 * const isPerson = is.ObjectOf({
 *   name: is.String,
 *   age: is.Number,
 *   address: is.OptionalOf(is.String),
 * });
 *
 * type Person = PredicateType<typeof isPerson>;
 * // Above is equivalent to the following type
 * // type Person = {
 * //   name: string;
 * //   age: number;
 * //   address: string | undefined;
 * // };
 */
export type PredicateType<P> = P extends Predicate<infer T> ? T : never;
