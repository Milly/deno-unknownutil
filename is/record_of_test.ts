import { assertEquals } from "@std/assert";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertType } from "@std/testing/types";
import type { Equal } from "../_testutil.ts";
import { is } from "./mod.ts";
import { isRecordOf } from "./record_of.ts";

Deno.test("isRecordOf<T>", async (t) => {
  await t.step("returns properly named predicate function", async (t) => {
    await assertSnapshot(t, isRecordOf(is.Number).name);
    await assertSnapshot(t, isRecordOf((_x): _x is string => false).name);
  });

  await t.step("returns true on T record", () => {
    assertEquals(isRecordOf(is.Number)({ a: 0 }), true);
    assertEquals(isRecordOf(is.String)({ a: "a" }), true);
    assertEquals(isRecordOf(is.Boolean)({ a: true }), true);
  });

  await t.step("returns false on non T record", () => {
    assertEquals(isRecordOf(is.String)({ a: 0 }), false);
    assertEquals(isRecordOf(is.Number)({ a: "a" }), false);
    assertEquals(isRecordOf(is.String)({ a: true }), false);
  });

  await t.step("predicated type is correct", () => {
    const a: unknown = { a: 0 };
    if (isRecordOf(is.Number)(a)) {
      assertType<Equal<typeof a, Record<PropertyKey, number>>>(true);
    }
  });
});

Deno.test("isRecordOf<T, K>", async (t) => {
  await t.step("returns properly named predicate function", async (t) => {
    await assertSnapshot(t, isRecordOf(is.Number, is.String).name);
    await assertSnapshot(
      t,
      isRecordOf((_x): _x is string => false, is.String).name,
    );
  });

  await t.step("returns true on T record", () => {
    assertEquals(isRecordOf(is.Number, is.String)({ a: 0 }), true);
    assertEquals(isRecordOf(is.String, is.String)({ a: "a" }), true);
    assertEquals(isRecordOf(is.Boolean, is.String)({ a: true }), true);
  });

  await t.step("returns false on non T record", () => {
    assertEquals(isRecordOf(is.String, is.String)({ a: 0 }), false);
    assertEquals(isRecordOf(is.Number, is.String)({ a: "a" }), false);
    assertEquals(isRecordOf(is.String, is.String)({ a: true }), false);
  });

  await t.step("returns false on non K record", () => {
    assertEquals(isRecordOf(is.Number, is.Number)({ a: 0 }), false);
    assertEquals(isRecordOf(is.String, is.Number)({ a: "a" }), false);
    assertEquals(isRecordOf(is.Boolean, is.Number)({ a: true }), false);
  });

  await t.step("predicated type is correct", () => {
    const a: unknown = { a: 0 };
    if (isRecordOf(is.Number, is.String)(a)) {
      assertType<Equal<typeof a, Record<string, number>>>(true);
    }
  });
});
