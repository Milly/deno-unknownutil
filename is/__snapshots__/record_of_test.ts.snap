export const snapshot = {};

snapshot[`isRecordOf<T> > returns properly named function 1`] = `"isRecordOf(isNumber, undefined)"`;

snapshot[`isRecordOf<T> > returns properly named function 2`] = `"isRecordOf((anonymous), undefined)"`;

snapshot[`isRecordOf<T, K> > returns properly named function 1`] = `"isRecordOf(isNumber, isString)"`;

snapshot[`isRecordOf<T, K> > returns properly named function 2`] = `"isRecordOf((anonymous), isString)"`;
